### PR TITLE
circulation: implement all the add_request actions

### PR DESCRIPTION
**Pull request by BadrAly**
Fixes problem when multiple pending requests are allowed for same
patron on same item.

Loans with state ITEM_IN_TRANSIT_TO_HOUSE are no longer blocking
new requests for same patron on same item.

Creates a method item_record_to_a_specific_loan_state to put an item
record in a specific status.

Reorganises the circulation units testing by creating a new directory
for circulation testings.

* Updates tests and fixtures accordingly.

Co-Authored-by: Aly Badr <aly.badr@rero.ch>

## Why are you opening this PR?

taiga_task: https://tree.taiga.io/project/rero21-reroils/task/1415

To implement the request actions defined by the group PO.
ref: https://hackmd.io/_3qU5KRiRZaLSYxaJrlsUA

## How to test?

- `run_tests.sh` 
- only code review at this moment.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

File: tests/api/circulation/test_item_api_views.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    res, _ = postdata(
+        client,
+        'api_item.librarian_request',
+        dict(
+            patron_pid=patron_martigny_no_email.pid
+        )
+    )
+    assert res.status_code == 500
+
+    # test successful when all parameters are given
```
Comment (blankoworld) [inline tests/api/circulation/test_item_api_views.py:58]: If the test is called "test_items_failed_actions", why not testing successful in another test?

Comment (BadrAly) [reply tests/api/circulation/test_item_api_views.py]: I moved this test from another file to here. Removed the successful action tests.

File: rero_ils/modules/loans/api.py [hunk 2]
```diff
@@ -57,6 +57,19 @@ class LoanAction(object):
     NO = 'no'
 
 
+class LoanState(object):
```
[the remark this answers is not in the document]
Comment (jma) [reply rero_ils/modules/loans/api.py:60]: @lauren-d is right: https://docs.python.org/3.6/library/enum.html#functional-api

File: rero_ils/modules/loans/api.py [hunk 3]
```diff
@@ -57,6 +57,19 @@ class LoanAction(object):
...
+    """Class holding all availabe circulation loan states."""
+
+    CREATED = 'CREATED'
+    PENDING = 'PENDING'
+    ITEM_IN_TRANSIT_FOR_PICKUP = 'ITEM_IN_TRANSIT_FOR_PICKUP'
```
Comment (jma) [inline rero_ils/modules/loans/api.py:65]: You can add translations if it is needed: `_('ITEM_IN_TRANSIT_FOR_PICKUP')`

File: rero_ils/modules/loans/api.py [hunk 4]
```diff
@@ -274,15 +287,35 @@ def create_notification(self, notification_type=None):
 
 def get_request_by_item_pid_by_patron_pid(item_pid, patron_pid):
     """Get pending, item_on_transit, item_at_desk loans for item, patron."""
+    filter_states = [
+        'PENDING',
```
Comment (jma) [inline rero_ils/modules/loans/api.py:291]: Why not using LoanState?

File: rero_ils/modules/loans/api.py [hunk 5]
```diff
@@ -274,15 +287,35 @@ def create_notification(self, notification_type=None):
 
 def get_request_by_item_pid_by_patron_pid(item_pid, patron_pid):
     """Get pending, item_on_transit, item_at_desk loans for item, patron."""
```
Comment (jma) [inline rero_ils/modules/loans/api.py:289]: params is missing

File: rero_ils/modules/loans/api.py (hunk 2 above)
Comment (jma) [inline rero_ils/modules/loans/api.py:60]: Seems a duplicate with item, loan statuses, no?

File: rero_ils/modules/loans/utils.py [hunk 6]
```diff
@@ -166,11 +167,17 @@ def can_be_requested(loan):
 
     if not loan.item_pid:
         raise Exception('Transaction on document is not implemented.')
-    # 1) Check if circulation_policy allows request
+
+    # 1) Check if there is already a loan for same patron+item
+    if get_any_loans_by_item_pid_by_patron_pid(
+            loan.get('item_pid').get('value'), loan.get('patron_pid')):
```
Comment (jma) [inline rero_ils/modules/loans/utils.py:173]: If the `item_pid` required for a loan, I would do `loan['item_pid']` else `loan.get('item_pid', {})`

Comment (jma) [reply rero_ils/modules/loans/utils.py:173]: The indentation is weird.

File: tests/api/circulation/test_item_api_views.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    # test fails for a request with no item given
+    res, _ = postdata(
+        client,
+        'api_item.librarian_request',
+        dict(
+            patron_pid=patron_martigny_no_email.pid
+        )
+    )
+    assert res.status_code == 500
```
Comment (jma) [inline tests/api/circulation/test_item_api_views.py:54]: bad idea! we never test 500 HTTP code as it is an `unexpected error`. In such case you need to handle properly the error: try/catch with the appropriate HTTP status code. You can perhaps find other cases in the test and fixed it with this PR.

File: tests/api/circulation/test_request_actions.py [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+    # test add_request_1_1 action
+    # item on_shelf (no current loan): PENDING loan does not exist
+    params = {
+        'patron_pid': patron_martigny_no_email.pid,
+        'transaction_location_pid': loc_public_martigny.pid,
+        'transaction_user_pid': librarian_martigny_no_email.pid,
+        'pickup_location_pid': loc_public_martigny.pid
+    }
+
+    item, loan = item_record_to_a_specific_loan_state(
```
Comment (jma) [inline tests/api/circulation/test_request_actions.py:43]: in the `tests/api` directory we have only the REST API tests. If you need to test the python API please put into `tests/ui/*/*_api.py`

**Pull request closed without merging**